homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,625 @@
+                trackid,
+                position,
+            )
+        except Exception as exc:
+            raise HomeAssistantError("cannot seek: %s" % exc) from exc
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return entity specific state attributes."""
+        if self.state == MediaPlayerState.OFF:
+            return {}
+        return {ATTR_PLAYBACK_RATE: self._attr_playback_rate}
+
+
+class MPRISCoordinator(
+    DataUpdateCoordinator[
+        dict[str, list[Union[mpris_pb2.MPRISPlayerUpdate, UNAVAILABLE]]]  # type: ignore[valid-type]
+    ]
+):
+    """The entity manager manages MPRIS media player entities.
+
+    This class is responsible for maintaining the known player entities
+    in sync with the state as reported by the server, as well as keeping
+    tabs of newly-appeared players and players that have gone.
+    """
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        mpris_client: hassmpris_client.AsyncMPRISClient,
+        async_add_entities: AddEntitiesCallback,
+    ) -> None:
+        """Initialize the entity manager.
+
+        Arguments:
+          hass: the HomeAssistant singleton
+          config_entry: the configuration entry associated with
+                        this component (or integration?)
+          mpris_client: the MPRIS client endpoint object
+          async_add_entities: callback to add entities async
+        """
+        super().__init__(hass, _LOGGER, name="MPRIS")
+        self.data: dict[str, list[Union[mpris_pb2.MPRISPlayerUpdate, UNAVAILABLE]]] = {}  # type: ignore[valid-type]
+        self.config_entry = config_entry
+        self.async_add_entities = async_add_entities
+        self._client = mpris_client
+        self._shutdown: asyncio.Future[bool] = asyncio.Future()
+        self._started = False
+        self._deferred_task: asyncio.Task | None = None
+
+    @property
+    def client(self) -> hassmpris_client.AsyncMPRISClient:
+        """Return the MPRIS client associated with this entity manager."""
+        return self._client
+
+    async def start(self) -> None:
+        """Start the entity manager as a separate task."""
+        self.hass.loop.create_task(self.run())
+
+    async def run(self) -> None:
+        """Run the entity manager."""
+        if self._started:
+            _LOGGER.debug("Thread already started")
+            return
+        self._started = True
+        _LOGGER.debug("Streaming updates started")
+        seen_excs: dict[Any, bool] = {}
+        while not self._shutdown.done():
+            try:
+                cycle_update_count = 0
+                try:
+                    async for _ in self._monitor_updates():
+                        cycle_update_count = cycle_update_count + 1
+                        seen_excs = {}
+                except Exception as exc:
+                    if self._shutdown.done():
```

Comment:
Recap:

1. `coordinator.stop()` is called from the unloaders.
2. `client.close()` is called immediately afterwards.

`self._monitor_updates()` only returns when the client is closed, and this raises an exception (that's just how GRPC works).

At that point in time, `self._shutdown.done()` will be True because `coordinator.stop()` was called *first*.  We use this as an indicator that this is a clean shutdown, and we suppress the exception coming back from `self.monitor_updates()`.

If you have a better idea on how to stop an async generator from GRPC, I am all ears.

Reply:
Please do feel free to reopen this thread if we need more clarity on what to do.  This specific part of the code has been massively abridged by reusing functionality of the `DataUpdateCoordinator` in the latest review address.